source/com.microsoft.tfs.client.common.ui/src/com/microsoft/tfs/client/common/ui/config/UITransportRequestHandler.java

```diff
@@ -278,18 +277,37 @@ public Status handleException(
         else if (exception instanceof UnauthorizedException && service.isPromptForCredentials()) {
             log.debug(" UnauthorizedException has been raised."); //$NON-NLS-1$
 
+            final Credentials usedCredentials = connectionInstanceData.getCredentials();
+            final java.net.URI serverUrl = connectionInstanceData.getServerURI();
+            final Header[] authHeaders = request.getPostMethod().getResponseHeaders("WWW-Authenticate"); //$NON-NLS-1$
+            boolean isHosted = true;
+            for (final Header header : authHeaders) {
+                if (header.getValue().equals("NTLM")) { //$NON-NLS-1$
+                    isHosted = false;
+                    break;
+                }
+                if (header.getValue().equals("Negotiate")) { //$NON-NLS-1$
+                    isHosted = false;
+                    break;
+                }
+                if (header.getValue().equals("Federated")) { //$NON-NLS-1$
+                    break;
+                }
+            }
+
             if (EnvironmentVariables.getBoolean(EnvironmentVariables.USE_OAUTH_LIBRARY, true)
-                && isPatCredentials(connectionInstanceData.getCredentials())) {
+                && usedCredentials != null
+                && new CachedCredentials(serverUrl, usedCredentials).isPatCredentials()
```

Comment:
Would it be better to swap this line with the line below so there's no needless object creation?